File: lib/agix-loop-sim/prng.mjs
// agix-loop-sim/prng — deterministic, seedable pseudo-random source.
//
// GUARDS: "reliability harness reads Math.random()" (loop-sim/RELIABILITY).
// The whole point of this harness is byte-for-byte reproducibility, so the
// simulation MUST NEVER touch Math.random() or any wall clock. Every draw
// comes from a mulberry32 stream keyed on a single integer seed: identical
// seed => identical stream, forever, across machines and Node versions.
//
// mulberry32 is a 32-bit generator with a full 2^32 period, excellent
// avalanche, and no external state — perfect for a CI-safe simulator.

/**
 * Raw mulberry32 core. Returns a function producing floats in [0, 1).
 * @param {number} seed  32-bit unsigned seed.
 */
export function mulberry32(seed) {
  let a = seed >>> 0;
  return function next() {
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

/**
 * A small typed API over a mulberry32 stream. All helpers are pure draws
 * from the same stream, so a fixed call ORDER on a fixed seed is fully
 * deterministic. Never introduce a draw whose presence depends on
 * wall-clock or environment state.
 *
 * @param {number} seed
 */
export function makePrng(seed) {
  const next = mulberry32(seed);
  const api = {
    /** Uniform float in [0, 1). */
    float() {
      return next();
    },
    /** Uniform float in [min, max). */
    range(min, max) {
      return min + (max - min) * next();
    },
    /** Uniform integer in [min, max] inclusive. */
    int(min, max) {
      return Math.floor(min + (max - min + 1) * next());
    },
    /** Bernoulli: true with probability p. */
    bool(p) {
      return next() < p;
    },
    /** Uniform pick from a non-empty array. */
    pick(arr) {
      return arr[Math.floor(next() * arr.length)];
    },
    /**
     * Weighted pick. `items` is [{ value, weight }] (weights > 0).
     * Stable: iterates in array order, so equal seeds pick equally.
     */
    weighted(items) {
      const total = items.reduce((s, it) => s + (it.weight ?? 1), 0);
      let r = next() * total;
      for (const it of items) {
        r -= it.weight ?? 1;
        if (r < 0) return it.value;
      }
      return items[items.length - 1].value;
    },
    /**
     * Standard-normal draw via Box–Muller (uses two stream draws).
     * Deterministic given the stream.
     */
    gaussian(mean = 0, sd = 1) {
      let u1 = next();
      const u2 = next();
      if (u1 < 1e-12) u1 = 1e-12; // avoid log(0)
      const mag = Math.sqrt(-2 * Math.log(u1));
      return mean + sd * mag * Math.cos(2 * Math.PI * u2);
    },
    /** Gaussian clamped to [lo, hi]. */
    clampedGaussian(mean, sd, lo, hi) {
      const v = api.gaussian(mean, sd);
      return Math.min(hi, Math.max(lo, v));
    },
  };
  return api;
}

File: core/kmstore/prng.go
// prng — deterministic, seedable pseudo-random source, ported byte-for-byte
// from lib/agix-loop-sim/prng.mjs (mulberry32). The whole point of a benchmark
// harness is reproducibility: identical seed => identical stream, forever,
// across machines and Go versions. This source NEVER touches math/rand or any
// wall clock, so corpus + query generation is fully deterministic.
//
// mulberry32 is a 32-bit generator with a full 2^32 period and excellent
// avalanche. The JS reference operates on int32/uint32 via `| 0`, `>>> n`, and
// Math.imul; the equivalent Go is plain uint32 arithmetic, which wraps mod 2^32
// with identical bit patterns — so this stream matches the Node premise test
// exactly for a given seed.
//
// Copyright 2026 Agix AI LLC. Apache-2.0.
package kmstore

import "math"

// prng is a mulberry32 stream keyed on a single 32-bit seed.
type prng struct{ a uint32 }

// newPRNG returns a stream seeded with the given 32-bit value.
func newPRNG(seed uint32) *prng { return &prng{a: seed} }

// next returns a uniform float in [0, 1). Mirrors the JS mulberry32 core:
//
//	a = (a + 0x6d2b79f5) | 0;
//	let t = Math.imul(a ^ (a >>> 15), 1 | a);
//	t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
//	return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
func (p *prng) next() float64 {
	p.a += 0x6d2b79f5
	t := (p.a ^ (p.a >> 15)) * (1 | p.a)
	t = (t + (t^(t>>7))*(61|t)) ^ t
	return float64(t^(t>>14)) / 4294967296.0
}

// intRange returns a uniform integer in [min, max] inclusive.
// Mirrors JS: Math.floor(min + (max - min + 1) * next()).
func (p *prng) intRange(min, max int) int {
	return min + int(math.Floor(float64(max-min+1)*p.next()))
}

// gaussian returns a standard-normal draw via Box–Muller (two stream draws),
// mirroring the JS gaussian(mean, sd).
func (p *prng) gaussian(mean, sd float64) float64 {
	u1 := p.next()
	u2 := p.next()
	if u1 < 1e-12 {
		u1 = 1e-12 // avoid log(0)
	}
	mag := math.Sqrt(-2 * math.Log(u1))
	return mean + sd*mag*math.Cos(2*math.Pi*u2)
}
